subprojects/publish/src/main/java/org/gradle/api/publish/internal/metadata/ModuleMetadataSpecBuilder.java

```diff
@@ -299,51 +294,27 @@ private Object attributeValueFor(Object value) {
 
     private ModuleMetadataSpec.DependencyCoordinates moduleDependencyCoordinatesFor(
         ModuleDependency dependency,
-        VariantVersionMappingStrategyInternal versionMappingStrategy
+        VariantDependencyResolver dependencyResolver
     ) {
-        String group = dependency.getGroup();
-        String name = dependency.getName();
-        String resolvedVersion = null;
-        if (versionMappingStrategy != null) {
-            ModuleVersionIdentifier resolvedVersionId = versionMappingStrategy.maybeResolveVersion(group, name, null);
-            if (resolvedVersionId != null) {
-                group = resolvedVersionId.getGroup();
-                name = resolvedVersionId.getName();
-                resolvedVersion = resolvedVersionId.getVersion();
-            }
-        }
+        VariantDependencyResolver.Coordinates coordinates = dependencyResolver.resolveComponentCoordinates(dependency);
         return new ModuleMetadataSpec.DependencyCoordinates(
-            group,
-            name,
-            versionFor(versionConstraintFor(dependency), resolvedVersion)
+            coordinates.getGroup(),
+            coordinates.getName(),
+            versionFor(versionConstraintFor(dependency), coordinates.getVersion())
         );
     }
 
     private ModuleMetadataSpec.DependencyCoordinates projectDependencyCoordinatesFor(
         ProjectDependency projectDependency,
-        VariantVersionMappingStrategyInternal versionMappingStrategy
+        VariantDependencyResolver variantDependencyResolver
     ) {
-        String resolvedVersion = null;
-        Path identityPath = ((ProjectDependencyInternal) projectDependency).getIdentityPath();
-        ModuleVersionIdentifier identifier = projectDependencyResolver.resolve(ModuleVersionIdentifier.class, identityPath);
-        if (versionMappingStrategy != null) {
-            ModuleVersionIdentifier resolved =
-                versionMappingStrategy.maybeResolveVersion(
-                    identifier.getGroup(),
-                    identifier.getName(),
-                    identityPath
-                );
-            if (resolved != null) {
-                identifier = resolved;
-                resolvedVersion = identifier.getVersion();
-            }
-        }
+        VariantDependencyResolver.Coordinates identifier = variantDependencyResolver.resolveComponentCoordinates(projectDependency);
         return new ModuleMetadataSpec.DependencyCoordinates(
```

Comment:
❓ Is there still a need for this `ModuleMetadataSpec.DependencyCoordinates` type?  Doesn't `VariantDependencyResolver.Coordinates` cover this?  Can it be consolidated with that type?

Reply:
I think the primary advantage is to have separate type hierarchies between the module metadata spec and the variant dependency resolver. These are two separate concepts and it would increase coupling of these two structures/interfaces with minimal benefit.

Reply:
I'm not necessarily saying the two coordinates are different concepts, but the type hierarchies that define the coordinate definitions are separate concepts. 

A ModuleMetadataSpec attempts to define a concrete definition of a GMM file. A VariantDependencyResolver attempts to determine the coordinates to publish some variant given the declared dependency. The metadata spec and variant dependency resolver are two separate concepts. One can define a metadata spec without resolving dependencies. They can also resolve declared dependencies without defining a metadata spec. 

There are at least some advantages to reducing coupling between these separate concepts. 

In any case, looking deeper, the `ModuleMetadataSpec.DependencyCoordinates` does not use a string-typed version field, but declares a `Version`, which can describe rich version metadata. The `Coordinates` interface uses a simple String.

Reply:
Wouldn't we just define multiple constructors and both `Version getVersion()` and `String getPlainVersion()` methods then.  

If this is a single concept, making it a single type seems better.  These types are static, and seem to be a self-contained concept that exists outside of either of these containing classes, so I don't see why the type hierarchies are a consideration here.

Reply:
There are advantages to keeping two types separate even though they _currently_ have similar interfaces. Particularly, it allows the two interfaces to evolve independently. 

`ModuleMetadataSpec` acts as _the_ definition of GMM. If we change this, we change the structure of GMM when it is published.

`PublishedCoordinates` is the return type of some unrelated code that happens to deal with coordinates. However, if these two independent code paths used the same type to represent their coordinates, and we wanted to make a modification to the resolving code, we would either:
1. Be forced to separate these two types again
2. Need to fundamentally change the definition of GMM

In particular, there are plans to further enhance the returned data from the coordinate resolvers to incorporate artifact data -- and thus they will need to include `classifier` and `type`. This will make the two types incompatible.

I really don't think we should merge these types. Even though they may seem similar, they represent two different concepts. Keeping these types separate allows us to evolve both GMM and coordinate resolution separately without affecting each other.

Reply:
Maybe another way of saying this is that the ModuleMetadataSpec is Gradle's external view of things and PublishedCoordinates is the projection of Gradle's model into Ivy/Maven?

Reply:
I like the projection terminology. 

`ModuleMetadataSpec` is the definition of a GMM file

`PublishedCoordinates` is the _closest representation_ of a Gradle dependency in the ivy and maven worlds